src/github.com/matrix-org/go-neb/services/google/google.go

```diff
@@ -0,0 +1,221 @@
+//
+// Example request:
+//   {
+//       "api_key": "AIzaSyA4FD39m9pN-hiYf2NRU9x9cOv5tekRDvM"
+//   }
+type Service struct {
+	types.DefaultService
+	// The Google API key to use when making HTTP requests to Google.
+	APIKey string `json:"api_key"`
+}
+
+// Commands supported:
+//    !google some search query without quotes
+// Responds with a suitable image into the same room as the command.
+func (s *Service) Commands(client *gomatrix.Client) []types.Command {
+	return []types.Command{
+		types.Command{
+			Path: []string{"google"},
+			Command: func(roomID, userID string, args []string) (interface{}, error) {
+				return s.cmdGoogle(client, roomID, userID, args)
+			},
+		},
+	}
+}
+
+// usageMessage returns a matrix TextMessage representation of the service usage
+func usageMessage() *gomatrix.TextMessage {
+	return &gomatrix.TextMessage{"m.notice",
+		`Usage: !google image image_search_text`}
+}
+
+func (s *Service) cmdGoogle(client *gomatrix.Client, roomID, userID string, args []string) (interface{}, error) {
+
+	if len(args) < 2 || args[0] != "image" {
+		return usageMessage(), nil
+	}
+	// Drop the search type (should currently always be "image")
+	args = args[1:]
+
+	// only 1 arg which is the text to search for.
+	querySentence := strings.Join(args, " ")
+
+	searchResult, err := s.text2imgGoogle(querySentence)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var imgURL = searchResult.Link
+	if imgURL == "" {
+		return gomatrix.TextMessage{
+			MsgType: "m.text.notice",
+			Body:    "No image found!",
+		}, nil
+	}
+
+	// FIXME -- Sometimes upload fails with a cryptic error - "msg=Upload request failed code=400 "
+	resUpload, err := client.UploadLink(imgURL)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to upload Google image to matrix: %s", err.Error())
+	}
+
+	img := searchResult.Image
+	return gomatrix.ImageMessage{
+		MsgType: "m.image",
+		Body:    querySentence,
+		URL:     resUpload.ContentURI,
+		Info: gomatrix.ImageInfo{
+			Height:   uint(math.Floor(img.Height)),
+			Width:    uint(math.Floor(img.Width)),
+			Mimetype: searchResult.Mime,
+		},
+	}, nil
+}
+
+// text2imgGoogle returns info about an image
+func (s *Service) text2imgGoogle(query string) (*googleSearchResult, error) {
+	log.Info("Searching Google for an image of a ", query)
+
+	u, err := url.Parse("https://www.googleapis.com/customsearch/v1")
+	if err != nil {
+		return nil, err
+	}
+
+	q := u.Query()
+	q.Set("q", query)            // String to search for
+	q.Set("num", "1")            // Just return 1 image result
+	q.Set("start", "1")          // No search result offset
+	q.Set("imgSize", "medium")   // Just search for medium size images
+	q.Set("searchType", "image") // Search for images
+	// q.set("fileType, "")                             // Any file format
+
+	var key = s.APIKey
+	if key == "" {
+		key = "AIzaSyA4FD39m9pN-hiYf2NRU9x9cOv5tekRDvM" // FIXME -- Should be instantiated from service config
```

Comment:
I've seen this referenced 3 times now: in `config.sample.yaml`, as an example request comment, and here. I'm guessing this is a beta key which Google provides for testing? As an aside, I would personally fail to register the service if no `APIKey` is supplied, rather than defaulting to this key if one isn't given.